testimony/__init__.py

```diff
@@ -92,190 +77,66 @@ def __init__(self, function_def, parent_class=None, testmodule=None):
-            line = line.rstrip()
-            # Sometimes there are double new line
-            # characters in the middle.  We need
-            # only one of those to print
-            line = line.replace('\n\n', '\n')
-            if len(line) > 0 and ':' in line:
-                tag, value = line.split(':', 1)
-                tag = tag.lower()
-                value = value.strip()
-                if tag in valid_tags:
-                    if tag == 'bz':
-                        value = [bz.strip() for bz in value.split(',')]
-                    expected_tags[tag] = value
-                else:
-                    unexpected_tags[tag] = value
-                    unexpected_lines.append(line)
-        return expected_tags, unexpected_tags, unexpected_lines
-
     def _parse_docstring(self):
-        """Parses the test docstring extracting expected values.
+        """Parse module, class and function docstrings.
 
-        If the expected tags is not spelled right they will not be parsed.
+        ``tokens`` and ``invalid_tokens`` attributes will be updated  with the
+        parsed values.
         """
         if self.docstring is None:
             return
 
-        # Create the contexts
-        module_docstring = ast.get_docstring(self.module_def)
-        if module_docstring and not isinstance(module_docstring, type(u'')):
-            module_docstring = module_docstring.decode('utf-8')
-        class_docstring = ast.get_docstring(self.parent_class_def)
-        if class_docstring and not isinstance(class_docstring, type(u'')):
-            class_docstring = class_docstring.decode('utf-8')
-        function_docstring = self.docstring
-        if function_docstring and not isinstance(
-                function_docstring, type(u'')):
-            function_docstring = function_docstring.decode('utf-8')
-        tags, unexpected_tags, _ = self._parse_tags(module_docstring)
-        class_tags, class_unexpected_tags, _ = self._parse_tags(
-            class_docstring)
-        function_tags, function_unexpected_tags, self.skipped_lines = (
-            self._parse_tags(function_docstring))
-
-        # Update context dictionaries
-        tags.update(class_tags)
-        tags.update(function_tags)
-        unexpected_tags.update(class_unexpected_tags)
-        unexpected_tags.update(function_unexpected_tags)
-
-        for tag, value in tags.items():
-            if tag == 'bz':
-                tag = 'bugs'
-            if tag == 'assert':
-                tag = 'assertion'
-            if tag == 'type':
-                tag = 'test_type'
-            setattr(self, tag, value)
-        self.unexpected_tags = unexpected_tags
+        # Parse module, class and function docstrings. Every loop update the
+        # already defined tokens and invalid_tokens. Doing in the mentioned
+        # order will ensure that function docstrings have more priority over
+        # class and module docstring respectively.
```

Comment:
Parse module, class and function docstrings.  Every loop updates the already defined tokens and invalid_tokens. The order of processing ensures that function docstring has more priority over class and module docstrings respectively.